cpu/nrf52/periph/spi.c

```diff
@@ -0,0 +1,258 @@
+ *
+ * @param bus   The SPI instance that is in use.
+ */
+static void _setup_workaround_for_ftpan_58(spi_t bus)
+{
+#ifdef ERRATA_SPI_SINGLE_BYTE_WORKAROUND
+    gpio_init_int(spi_config[bus].sclk, GPIO_OUT, GPIO_BOTH,
+                  spi_gpio_handler, (void *)bus);
+    gpio_irq_disable(spi_config[bus].sclk);
+    uint8_t channel = gpio_int_get_exti(spi_config[bus].sclk);
+    assert(channel != 0xff);
+
+    // Stop the spim instance when SCK toggles.
+    NRF_PPI->CH[spi_config[bus].ppi].EEP =
+        (uint32_t)&NRF_GPIOTE->EVENTS_IN[channel];
+    NRF_PPI->CH[spi_config[bus].ppi].TEP = (uint32_t)&dev(bus)->TASKS_STOP;
+#else
+    (void)bus;
+#endif
+}
+
+static void _enable_workaround(spi_t bus)
+{
+#ifdef ERRATA_SPI_SINGLE_BYTE_WORKAROUND
+    /**
+     * The spim instance cannot be stopped mid-byte, so it will finish
+     * transmitting the first byte and then stop. Effectively ensuring
+     * that only 1 byte is transmitted.
+     */
+    NRF_PPI->CHENSET = 1U << spi_config[bus].ppi;
+    gpio_irq_enable(spi_config[bus].sclk);
+#else
+    (void)bus;
+#endif
+}
+
+static void _clear_workaround(spi_t bus)
+{
+#ifdef ERRATA_SPI_SINGLE_BYTE_WORKAROUND
+    NRF_PPI->CHENCLR = 1U << spi_config[bus].ppi;
+#else
+    (void)bus;
+#endif
+}
+
+void spi_init(spi_t bus)
+{
+    assert(bus < SPI_NUMOF);
+
+    /* initialize mutex */
+    mutex_init(&locks[bus]);
+    mutex_init(&busy[bus]);
+    mutex_lock(&busy[bus]);
+    /* initialize pins */
+    spi_init_pins(bus);
+}
+
+void spi_init_pins(spi_t bus)
+{
+    gpio_init(spi_config[bus].sclk, GPIO_OUT);
+    gpio_init(spi_config[bus].mosi, GPIO_OUT);
+    gpio_init(spi_config[bus].miso, GPIO_IN);
+    /* select pins for the SPI device */
+    SPI_SCKSEL = spi_config[bus].sclk;
+    SPI_MOSISEL = spi_config[bus].mosi;
+    SPI_MISOSEL = spi_config[bus].miso;
+    _setup_workaround_for_ftpan_58(bus);
+    spi_twi_irq_register_spi(dev(bus), spi_isr_handler, (void *)bus);
+}
+
+int spi_acquire(spi_t bus, spi_cs_t cs, spi_mode_t mode, spi_clk_t clk)
+{
+    (void)cs;
+
+    mutex_lock(&locks[bus]);
+    /* configure bus */
+    dev(bus)->CONFIG = mode;
+    dev(bus)->FREQUENCY = clk;
+    /* enable the bus */
+    dev(bus)->ENABLE = SPIM_ENABLE_ENABLE_Enabled;
+
+    return SPI_OK;
+}
+
+void spi_release(spi_t bus)
+{
+    /* power off everything */
+    dev(bus)->ENABLE = 0;
+    mutex_unlock(&locks[bus]);
+}
+
+static size_t _transfer(spi_t bus, const uint8_t *out_buf, uint8_t *in_buf,
+                      size_t remaining_len)
+{
+    uint8_t transfer_len = remaining_len > UINT8_MAX ? UINT8_MAX : remaining_len;
+    const uint8_t *out_mbuf = out_buf;
+
+    /**
+     * Copy the out buffer in case it resides in flash, EasyDMA only works from
+     * RAM
+     */
+    if (out_buf && !_in_ram(out_buf)) {
+        /* The SPI MBUF can be smaller than UINT8_MAX */
+        transfer_len = transfer_len > SPI_MBUF_SIZE ?
+            SPI_MBUF_SIZE : transfer_len;
```

Comment:
```suggestion
                       SPI_MBUF_SIZE : transfer_len;
```

minor nit.

```suggestion
        transfer_len = transfer_len > SPI_MBUF_SIZE
                     ? SPI_MBUF_SIZE : transfer_len;
```
Is visually even more pleasing, but that's personal taste :wink: